File: test/data/source/sync/device_sync_report_store_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:openvitals/data/source/sync/device_sync_report_store.dart';
import 'package:openvitals/data/source/sync/sync_report.dart';

void main() {
  group('buildSyncReportText', () {
    final at = DateTime.utc(2026, 1, 20, 14, 30);

    test('renders the summary and per-type lines for a completed sync', () {
      final report = SyncReport(
        completed: true,
        peerDeviceName: 'Galaxy S23',
        negotiatedTypes: const ['WeightRecord'],
        itemsSent: 10,
        itemsReceived: 8,
        imported: 6,
        duplicateSkipped: 2,
        typeSummaries: const [
          SyncTypeSummary(recordType: 'WeightRecord', received: 8, imported: 6, duplicateSkipped: 2),
        ],
      );
      final text = buildSyncReportText(report, generatedAt: at);
      expect(text, contains('Status: completed'));
      expect(text, contains('Peer: Galaxy S23'));
      expect(text, contains('Imported: 6'));
      expect(text, contains('Already had (skipped): 2'));
      expect(text, contains('WeightRecord: received 8, imported 6, skipped 2'));
      expect(text, contains('2026-01-20T14:30'));
    });

    test('renders the abort reason for an aborted sync', () {
      final report = SyncReport(
        completed: false,
        peerDeviceName: 'unknown',
        negotiatedTypes: const [],
        itemsSent: 0,
        itemsReceived: 0,
        imported: 0,
        duplicateSkipped: 0,
        typeSummaries: const [],
        abortReason: 'pairing code did not match',
      );
      final text = buildSyncReportText(report, generatedAt: at);
      expect(text, contains('Status: aborted'));
      expect(text, contains('Reason: pairing code did not match'));
      expect(text, contains('(none)'));
    });
  });

  group('DeviceSyncReportStore', () {
    test('round-trips a report through a file', () async {
      final dir = await Directory.systemTemp.createTemp('ovsync');
      final store = DeviceSyncReportStore(directoryResolver: () async => dir);
      expect(await store.readReport(), '');
      await store.writeReport('hello report');
      expect(await store.readReport(), 'hello report');
      await dir.delete(recursive: true);
    });
  });
}

File: app/src/test/kotlin/tech/mmarca/openvitals/data/source/sync/DeviceSyncReportStoreTest.kt
package tech.mmarca.openvitals.data.source.sync

import android.content.Context
import io.mockk.every
import io.mockk.mockk
import java.io.File
import java.time.Instant
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import tech.mmarca.openvitals.data.repository.TestDispatcherProvider
import tech.mmarca.openvitals.features.devicesync.protocol.SyncReport
import tech.mmarca.openvitals.features.devicesync.protocol.SyncTypeSummary
import tech.mmarca.openvitals.features.devicesync.protocol.buildSyncReportText
import tech.mmarca.openvitals.features.devicesync.store.DeviceSyncReportStore

/**
 * Port of the Flutter `device_sync_report_store_test.dart` suite. The Dart
 * store takes a directory resolver; the Kotlin one derives its directory from
 * `Context.filesDir`, so the temp directory is injected through a mocked
 * context instead.
 */
class DeviceSyncReportStoreTest {

    @get:Rule
    val temporaryFolder = TemporaryFolder()

    private val at: Instant = Instant.parse("2026-01-20T14:30:00Z")

    @Test
    fun `renders the summary and per-type lines for a completed sync`() {
        val report = SyncReport(
            completed = true,
            peerDeviceName = "Galaxy S23",
            negotiatedTypes = listOf("WeightRecord"),
            itemsSent = 10,
            itemsReceived = 8,
            imported = 6,
            duplicateSkipped = 2,
            typeSummaries = listOf(
                SyncTypeSummary(
                    recordType = "WeightRecord",
                    received = 8,
                    imported = 6,
                    duplicateSkipped = 2,
                ),
            ),
        )

        val text = buildSyncReportText(report, generatedAt = at)

        assertTrue(text.contains("Status: completed"))
        assertTrue(text.contains("Peer: Galaxy S23"))
        assertTrue(text.contains("Imported: 6"))
        assertTrue(text.contains("Already had (skipped): 2"))
        assertTrue(text.contains("WeightRecord: received 8, imported 6, skipped 2"))
        assertTrue(text.contains("2026-01-20T14:30"))
    }

    @Test
    fun `renders the abort reason for an aborted sync`() {
        val report = SyncReport(
            completed = false,
            peerDeviceName = "unknown",
            negotiatedTypes = emptyList(),
            itemsSent = 0,
            itemsReceived = 0,
            imported = 0,
            duplicateSkipped = 0,
            typeSummaries = emptyList(),
            abortReason = "pairing code did not match",
        )

        val text = buildSyncReportText(report, generatedAt = at)

        assertTrue(text.contains("Status: aborted"))
        assertTrue(text.contains("Reason: pairing code did not match"))
        assertTrue(text.contains("(none)"))
    }

    @Test
    fun `round-trips a report through a file`() = runTest {
        val filesDir: File = temporaryFolder.newFolder("files")
        val context = mockk<Context> {
            every { this@mockk.filesDir } returns filesDir
        }
        val store = DeviceSyncReportStore(context, TestDispatcherProvider)

        assertEquals("", store.readReport())
        store.writeReport("hello report")
        assertEquals("hello report", store.readReport())
    }
}
